runtime/main/src/main/java/org/qbicc/runtime/main/Once.java

```diff
@@ -0,0 +1,60 @@
+package org.qbicc.runtime.main;
+
+/**
+ * A one-time operation.
+ */
+public final class Once {
+
+    private final Runnable target;
+    private final Object lock = new Object();
+    private Throwable thrown;
+    private volatile boolean done;
+
+    /**
+     * Construct a new instance.
+     *
+     * @param target the target to run
+     */
+    public Once(Runnable target) {
+        this.target = target;
+    }
+
+    /**
+     * Execute the one-time task.
+     * If it throws an exception, the thrown object propagates to the caller.
+     * If the task was previously run successfully, this method does nothing.
+     * If the task was previously run but failed, the same thrown object is propagated to the caller.
+     * If the task is in the process of being run by the current thread, this method returns without doing anything.
+     * If the task is in the process of being run by some other thread, the calling thread will block until the task completes.
+     *
+     * @throws Throwable the object thrown by the target task
+     */
+    public void run() throws Throwable {
+        boolean done = this.done;
+        if (! done) {
+            if (Thread.holdsLock(lock)) {
+                // we're already running it
+                return;
+            }
+            synchronized (lock) {
+                done = this.done;
+                if (! done) try {
+                    Runnable target = this.target;
+                    if (target != null) {
+                        target.run();
+                    }
+                    return;
+                } catch (Throwable t) {
+                    thrown = t;
+                    // volatile write publishes thrown
+                    this.done = true;
+                    throw t;
+                }
+            }
+        }
+        Throwable thrown = this.thrown;
+        if (thrown != null) {
+            throw thrown;
+        }
```

Comment:
Can we move this code into the synchronized block?  The common case will be that the `Once` operation has already run and succeeded so coding this so the fast path only does one check would be better.

If we don't set `this.done = true` in the catch block, and moved this check to the start of the `synchronized` block, failed cases would be slower but overall runtime would be better